### DISCUSSION TITLE
How do I compile a shared library to be used with ctypes on mesonpy?

### DISCUSSION THREAD

**Discussion by peppedilillo** [mesonbuild/meson-python#573] [Q&A] [answered]
Hello.

Suppose I have pip-installed meson, mesonpy and ninja. I have a project structured like this:

``` 
source/
source/include/
source/include/mylib.h
source/mylib.c
source/wrapper.py
meson.build
pyproject.toml
```

With the following `meson.build`:
```
project('mylib', 'c')
add_global_arguments('-D NDEBUG', language : 'c')
incdir = include_directories('source/include')
sources = ['source/mylib.c']
shared_library('mylib', sources, include_directories : incdir)
```

I can build a file `build/myliblib.dylib` using `meson build` and then moving to `/build` to call `ninja`.
Now  I can use `lib = ctypes.CDLL("./builddir/myliblib.dylib")` to load the library in `wrapper.py` and do my things.

Obviously having to compile the shared library directly sucks, and I'd like the user to just `pip install .` and have the shared library be build in a defined place so that the ctypes wrapper works out of the box.

How am I supposed to do this with `mesonpy`? 
I don't want to write an extension for a moment, nor to use cython, but all the examples i can find seems to do either of those.

**Comment by rgommers** [comment]
> and I'd like the user to just `pip install .` and have the shared library be build in a defined place so that the ctypes wrapper works out of the box.

You need a Python package to make `pip install .` work, so it looks like you need a name (I assume it won't be `source`). So you want it to end up inside site-packages like:
```
mypkg/wrapper.py
mypkg/mylib.so
```

To do that, you need something like this:
```meson
py = import('python').find_installation(pure: false)

py.install_sources(
    ['source/wrapper.py'],
    subdir: 'mypkg'
)

shared_library('mylib'
    ['source/mylib.c'],
    include_directories : ['source/include'],
    install: true,
    install_dir: py.get_install_dir() / 'mypkg',
)
```

The `install_dir` part is what you were missing I guess? Note that your shared library should be inside your Python package; you cannot put it in the standard library directory and then expect `wrapper.py` to find it.

**Reply by dnicolodi** [reply]
This is one way of doing it. The other way is to let Meson install the library in the default location and let meson-python do its magic. meson-python installs shared libraries in a package dependent location alongside the package and you can instruct ctypes to load it from there.

The advantage of the solution described by @rgommers is that the installation location of the library does not depend on whether the package is installed as a wheel via `pip` or simply with the `meson setup`, `ninja`, `meson install` sequence. The advantage of letting meson-python install the shared library is that when not installed as a Python package, the library is in the standard location. However, in the latter case, your Python package needs to lock for the library in two places.

Which mechanism to use depends on your requirements. If users always interact with the library via the Python wrapper, the solution described by @rgommers is the most straightforward, if users may want to link your library into their code, a conventional installation location when the package is installed not as a Python wheel is more convenient. Unfortunately, at the moment a solution that works well for both uses does not exist.

**Comment by peppedilillo** [comment]
dear Ralf and Daniele,

thank you very much for your help, I truly appreciate. 
I tried replicating Ralf's solution. It seems ok to me since I don't expect my users to use the library outside the code I'm providing. 
I built a project as you suggested and pip-installed it. 
The installation gets to the end without errors, this is the output it produces:

```shell
(testenv) peppedilillo@pepbook meson-test % pip install .
Processing /Users/peppedilillo/Library/CloudStorage/Dropbox/Progetti/meson-test
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: mypkg
  Building wheel for mypkg (pyproject.toml) ... done
  Created wheel for mypkg: filename=mypkg-0.0.1-cp311-cp311-macosx_13_0_arm64.whl size=3841 sha256=37c69cd5d2b53c49a0470730fc04be4c3d2ce130dc36e66dd6cce4610a118b64
  Stored in directory: /Users/peppedilillo/Library/Caches/pip/wheels/2e/c7/6c/98b3ab681413a1de05be5bda2122cd94f19923c3e95b1b66e2
Successfully built mypkg
Installing collected packages: mypkg
  Attempting uninstall: mypkg
    Found existing installation: mypkg 0.0.1
    Uninstalling mypkg-0.0.1:
      Successfully uninstalled mypkg-0.0.1
Successfully installed mypkg-0.0.1
```

However I can't find the shared library where I'm expecting it to be, i.e. in the directory`mypkg/`!
I tried looking in my virtual environment's python folder but I could not find `mylib.so` there neither.
I have created a MWE for this, at [this repo](https://github.com/peppedilillo/meson-test/tree/main).

My `meson.build` is the following:

```
project('MyPkg', 'c')
add_global_arguments('-D NDEBUG', language : 'c')

py = import('python').find_installation(pure: false)

py.install_sources(
    ['mypkg/wrapper.py'],
    subdir: 'mypkg'
)

incdir = include_directories('source/include')
sources = ['source/mylib.c']
shared_library('mylib',
    sources,
    include_directories : incdir,
    install: true,
    install_dir: py.get_install_dir() / 'mypkg'
)
```

Thank you once again. Not only you are making a lot of science possible with numpy, scipy, meson-python and the alike.. You are giving many scientists something to be passionated about!!! Bests.

**Reply by dnicolodi** [reply]
It works exactly as expected for me. 
```console
$ git clone git@github.com:peppedilillo/meson-test.git
...
$ cd meson-test
$ python3.12 -m venv .
$ bin/python -m pip install .
...
$ find lib -name libmylib.dylib
lib/python3.12/site-packages/mypkg/libmylib.dylib
```
This is on macOS where the shared library filename suffix is `.dylib`. On Linux it is `.so` and `.dll` on Windows, obviously.

Where do you expect the shared library to be installed?

**Reply by peppedilillo** [reply]
Oh god here it is. I was expecting It to be in `meson-test/mypkg`. The problem seems to be due to having installed into a conda environment, so the library gets installed into my buried environment folder, making It less obvious where to find it:

```
(testenv) peppedilillo@pepbook meson-test % which python
/Users/peppedilillo/anaconda3/envs/testenv/bin/python
cd /Users/peppedilillo/anaconda3/envs/testenv
find lib -name libmylib.dylib
```
 
Which results in: 
```
lib/python3.11/site-packages/mypkg/libmylib.dylib
```

Huge step forward I guess, sorry for the noobery. 
One more question, what if I wanted to place the shared library in the `source` folder of my project, or even in the folder `mypkg`, where `wrapper.py` lives? 
This is to avoid having to call `sys.executable` to locate the python installation folder, then the library. But maybe there are better ways to find the shared library at runtime I am missing? Because It feels like `lib` is its right place. I don't know.

To install the library at `meson-test/mypkg`, I tried defining the root folder with `meson.source_root()`

```
project('MyPkg', 'c')
add_global_arguments('-D NDEBUG', language : 'c')

root = meson.source_root()

py = import('python').find_installation(pure: false)

py.install_sources(
    ['mypkg/wrapper.py'],
    subdir: 'mypkg'
)

incdir = include_directories('source/include')
sources = ['source/mylib.c']
shared_library('mylib',
    sources,
    include_directories : incdir,
    install: true,
    install_dir: root / 'mypkg'
)
```

but installation fails with:

```shell
('\x1b[31m',)meson-python: error: Could not map installation path to an equivalent wheel directory: '/Users/peppedilillo/Library/CloudStorage/Dropbox/Progetti/meson-test/source/libmylib.dylib'
      [end of output]
```


Thank you.

**Reply by rgommers** [reply]
> To install the library at `meson-test/mypkg`, I tries defining a root with `meson.source_root()`

Please don't do that, that won't work. I assume you want to do this for ease of development? What you're looking for then is an editable install: https://mesonbuild.com/meson-python/how-to-guides/editable-installs.html

**Reply by peppedilillo** [reply]
This is not really for ease of development but rather for making it easy and reliable loading the library with ctypes. 

In `wrapper.py` if I try to load the library with:

```python
lib = ctypes.CDLL("libmylib.dylib")
```

I get an `OsError` because python doesn't look for my shared library into `meson-test/lib/python3.10/site-packages/` , where it is actually installed.

```shell
OSError: dlopen(libmylib.dylib, 0x0006): tried: 'libmylib.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibmylib.dylib' (no such file), '/Users/peppedilillo/anaconda3/envs/meson-test/lib/python3.10/lib-dynload/../../libmylib.dylib' (no such file), '/Users/peppedilillo/anaconda3/envs/meson-test/bin/../lib/libmylib.dylib' (no such file), '/usr/lib/libmylib.dylib' (no such file, not in dyld cache), 'libmylib.dylib' (no such file), '/usr/local/lib/libmylib.dylib' (no such file), '/usr/lib/libmylib.dylib' (no such file, not in dyld cache)
```

If `libmylib.dylib` was installed in  the directory `meson-test/mypkg`, which also hosts `wrapper.py`, the library would have loaded.

A workaround is to just copypaste the library's path:
```
lib = ctypes.CDLL("/Users/peppedilillo/anaconda3/envs/meson-test/lib/python3.10/site-packages/mypkg/libmylib.dylib")
```
This works, but it kills portability. One could come up with tricks to get around this (for example messing with `Path` or using `sys.executable` to get the python installation folder) but it looks like a very fragile approach. 

I tried looking for the library path using `ctypes.util.find_library` but I could not get it, at least on MacOs.

Maybe am I missing something?


PS: I'm going to use editable installs from now on, it's great!

**Reply by rgommers** [reply, marked as answer]
> One could come up with tricks to get around this (for example messing with `Path` or using `sys.executable` to get the python installation folder)

You'll want to look into how regular Python packages work. You should have `wrapper.py` and an `__init__.py` right next to your shared library if you install to site-packages. And then you use it via an `import` statement, not via running `python wrapper.py`.

**Reply by peppedilillo** [reply]
I finally understand 🥹. I will research into packaging to get a firmer understanding of how it works. once done I will write a recap on this and post it somewhere. maybe it will help someone with the same misunderstandings in the future, and give meson-py something back for the insights i got through this thread. thank you again and all the best!!
